stdlib/public/core/Demangle.swift

```diff
@@ -0,0 +1,145 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2019 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SwiftShims
+
+/// Represents the potential return types from a call to demangle.
+public enum DemangleResult: Equatable {
+  /// The demangle completed successfully.
+  case success
+
+  /// The demangle resulted in truncating the result. The payload value is the
+  /// number of bytes necessary for a full demangle.
+  case truncated(Int)
+
+  /// The passed Swift mangled symbol was invalid.
+  case invalidSymbol
+}
+
+/// Given a mangled Swift symbol, demangle it into a human readable format.
+///
+/// Valid Swift symbols begin with the following prefixes:
+///   ┌─────────────────────╥────────┐
+///   │ Swift Version       ║        │
+///   ╞═════════════════════╬════════╡
+///   │ Swift 3 and below   ║   _T   │
+///   ├─────────────────────╫────────┤
+///   │ Swift 4             ║  _T0   │
+///   ├─────────────────────╫────────┤
+///   │ Swift 4.x           ║   $S   │
+///   ├─────────────────────╫────────┤
+///   │ Swift 5+            ║   $s   │
+///   └─────────────────────╨────────┘
+///
+/// - Parameters:
+///   - mangledName: A mangled Swift symbol.
+/// - Returns: A human readable demangled Swift symbol.
+public func demangle(_ mangledName: String) -> String? {
+  return mangledName.utf8CString.withUnsafeBufferPointer {
+    let demangledPtr = _swift_stdlib_demangle(
+      /* mangledName */ $0.baseAddress,
+      /* mangledNameLength */ $0.count - 1,
+      /* outputBuffer */ nil,
+      /* outputBufferSize */ nil,
+      /* flags */ 0
+    )
+
+    guard demangledPtr != nil else {
+      return nil
+    }
+
+    let demangledName = String(cString: demangledPtr!)
+    _swift_stdlib_free(demangledPtr!)
+    return demangledName
+  }
+}
+
+/// Given a mangled Swift symbol, demangle it into a human readable format.
+///
+/// Valid Swift symbols begin with the following prefixes:
+///   ┌─────────────────────╥────────┐
+///   │ Swift Version       ║        │
+///   ╞═════════════════════╬════════╡
+///   │ Swift 3 and below   ║   _T   │
+///   ├─────────────────────╫────────┤
+///   │ Swift 4             ║  _T0   │
+///   ├─────────────────────╫────────┤
+///   │ Swift 4.x           ║   $S   │
+///   ├─────────────────────╫────────┤
+///   │ Swift 5+            ║   $s   │
+///   └─────────────────────╨────────┘
+///
+/// - Parameters:
+///   - mangledNameBuffer: A buffer pointer pointing to a null-terminated C
```

Comment:
If `mangledNameBuffer` is always null-terminated then you could change it to `mangledName: UnsafePointer<CChar>`, so that `String` arguments can be implicitly converted.

Reply:
We already have the String argument version, so I'm unsure why supporting a pointer version for implicit conversion is necessary.

Reply:
I'm suggesting that:

* either `mangledNameBuffer: UnsafeBufferPointer<Int8>` doesn't need to be null-terminated;

* or `mangledName: UnsafePointer<Int8>` would make the third `demangle` function unnecessary.

On the other hand, John McCall wants to [deprecate the implicit conversions](https://forums.swift.org/t/revisiting-the-pointer-conversions/19363).

By the way, are `Int8` or `UInt8` elements more suitable for UTF-8 string parameters?

Reply:
If we use your suggestion about using `StringRef(mangledName, mangledNameLength))` mangledNameBuffer doesn't need to be null terminated. However, we must commit to either the name buffer being null terminated (because we calculate length with nameBuffer.count - 1), or must not be null terminated (nameBuffer.count).

As for `Int8` or `UInt8`, `CChar` maps to `Int8` and the string functions `withCString` and `utf8CString.withUnsafeBufferPointer`, etc all have `Int8` element types. `withUTF8` on the other hand uses `UInt8`, so I'm unsure. I think going with `Int8` is the better call because of the `CChar` typealias.

Reply:
`CChar` is either signed or unsigned, depending on the platform.